File: android/core/src/main/kotlin/pes/core/TimeUtil.kt
/**
 * Time handling: UTC formatting, DST resolution, local days, quiet zones
 * (spec §4). Mirrors `pes/core/timeutil.py`.
 *
 * java.time's defaults match the spec's DST rules: constructing a
 * ZonedDateTime at a nonexistent local time shifts it forward by the gap
 * length (the pre-transition offset applied), and an ambiguous local time
 * resolves to the earlier offset (first occurrence).
 */
package pes.core

import java.time.DayOfWeek
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.ZoneId
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import kotlinx.serialization.json.JsonObject

val WEEKDAYS = listOf("mon", "tue", "wed", "thu", "fri", "sat", "sun")

private val UTC_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'")
private val LOCAL_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss")

/** ISO-8601 `...Z` string -> epoch seconds. */
fun parseUtc(s: String): Long =
    LocalDateTime.parse(s, UTC_FORMAT).toEpochSecond(ZoneOffset.UTC)

/** Epoch seconds -> ISO-8601 `...Z` string, second resolution. */
fun fmtUtc(epoch: Long): String =
    LocalDateTime.ofEpochSecond(epoch, 0, ZoneOffset.UTC).format(UTC_FORMAT)

/** Epoch seconds -> local wall-clock ISO string without offset (exports). */
fun fmtLocal(epoch: Long, zone: ZoneId): String =
    Instant.ofEpochSecond(epoch).atZone(zone).format(LOCAL_FORMAT)

/**
 * Resolve a wall-clock time on a local day to (epochSeconds, isGap).
 * Nonexistent times resolve with the pre-transition offset (isGap = true);
 * ambiguous times resolve to the first occurrence.
 */
fun resolveLocal(day: LocalDate, hh: Int, mm: Int, zone: ZoneId): Pair<Long, Boolean> {
    val naive = LocalDateTime.of(day, LocalTime.of(hh, mm))
    val gap = zone.rules.getValidOffsets(naive).isEmpty()
    return Pair(naive.atZone(zone).toEpochSecond(), gap)
}

/** UTC bounds [start, end) of a local calendar day (23, 24, or 25 h). */
fun localDayBounds(day: LocalDate, zone: ZoneId): Pair<Long, Long> {
    val (start, _) = resolveLocal(day, 0, 0, zone)
    val (end, _) = resolveLocal(day.plusDays(1), 0, 0, zone)
    return Pair(start, end)
}

fun parseHhmm(s: String): Pair<Int, Int> {
    val (hh, mm) = s.split(":")
    return Pair(hh.toInt(), mm.toInt())
}

fun weekdayName(d: LocalDate): String = WEEKDAYS[d.dayOfWeek.value - DayOfWeek.MONDAY.value]

/**
 * True if the instant falls in any quiet zone, evaluated on wall-clock local
 * time (spec §4). Windows are half-open [from, to). A midnight-wrapping
 * window (from > to) starts on each listed day and runs into the next day.
 */
fun inQuietZone(epoch: Long, zones: List<JsonObject>, zone: ZoneId): Boolean {
    val local = Instant.ofEpochSecond(epoch).atZone(zone)
    val t = local.toLocalTime()
    val today = weekdayName(local.toLocalDate())
    val yesterday = weekdayName(local.toLocalDate().minusDays(1))
    for (z in zones) {
        val days = z.optStringList("days")?.toSet() ?: WEEKDAYS.toSet()
        val (fh, fm) = parseHhmm(z.str("from"))
        val (th, tm) = parseHhmm(z.str("to"))
        val frm = LocalTime.of(fh, fm)
        val to = LocalTime.of(th, tm)
        if (frm < to) {
            if (today in days && t >= frm && t < to) return true
        } else { // wraps midnight
            if ((today in days && t >= frm) || (yesterday in days && t < to)) return true
        }
    }
    return false
}

fun utcDayOf(epoch: Long): LocalDate = LocalDate.ofInstant(Instant.ofEpochSecond(epoch), ZoneOffset.UTC)

fun utcMidnight(day: LocalDate): Long = day.atStartOfDay(ZoneOffset.UTC).toEpochSecond()

File: desktop/pes/core/timeutil.py
"""Time handling: UTC formatting, DST resolution, local days, quiet zones (spec §4).

All elapsed-time arithmetic is in UTC epoch seconds (ints). Wall-clock
specifications are converted to UTC instants with the configured timezone
under the spec's DST rules:

- nonexistent local time (spring-forward gap) -> the instant the wall clock
  would have shown it under the pre-transition offset, flagged as a gap;
- duplicated local time (fall-back) -> the first occurrence.
"""

from __future__ import annotations

from datetime import UTC, date, datetime, time, timedelta
from zoneinfo import ZoneInfo

WEEKDAYS = ("mon", "tue", "wed", "thu", "fri", "sat", "sun")


def parse_utc(s: str) -> int:
    """ISO-8601 ``...Z`` string -> epoch seconds."""
    return int(datetime.strptime(s, "%Y-%m-%dT%H:%M:%SZ").replace(tzinfo=UTC).timestamp())


def fmt_utc(epoch: int) -> str:
    """Epoch seconds -> ISO-8601 ``...Z`` string, second resolution."""
    return datetime.fromtimestamp(epoch, UTC).strftime("%Y-%m-%dT%H:%M:%SZ")


def fmt_local(epoch: int, tz: ZoneInfo) -> str:
    """Epoch seconds -> local wall-clock ISO string without offset (exports)."""
    return datetime.fromtimestamp(epoch, tz).strftime("%Y-%m-%dT%H:%M:%S")


def resolve_local(day: date, hh: int, mm: int, tz: ZoneInfo) -> tuple[int, bool]:
    """Resolve a wall-clock time on a local day to (epoch_seconds, is_gap).

    ``fold=0`` gives PEP 495 semantics matching the spec: nonexistent times
    resolve with the pre-transition offset (is_gap=True); ambiguous times
    resolve to the first occurrence.
    """
    naive = datetime.combine(day, time(hh, mm))
    utc = naive.replace(tzinfo=tz, fold=0).astimezone(UTC)
    round_trip = utc.astimezone(tz).replace(tzinfo=None)
    return int(utc.timestamp()), round_trip != naive


def local_day_bounds(day: date, tz: ZoneInfo) -> tuple[int, int]:
    """UTC bounds [start, end) of a local calendar day (23, 24, or 25 h)."""
    start, _ = resolve_local(day, 0, 0, tz)
    end, _ = resolve_local(day + timedelta(days=1), 0, 0, tz)
    return start, end


def parse_hhmm(s: str) -> tuple[int, int]:
    hh, mm = s.split(":")
    return int(hh), int(mm)


def weekday_name(d: date) -> str:
    return WEEKDAYS[d.weekday()]


def in_quiet_zone(epoch: int, zones: list[dict], tz: ZoneInfo) -> bool:
    """True if the instant falls in any quiet zone, evaluated on wall-clock
    local time (spec §4). Windows are half-open [from, to). A midnight-wrapping
    window (from > to) starts on each listed day and runs into the next day.
    """
    local = datetime.fromtimestamp(epoch, tz)
    t = local.time()
    today = weekday_name(local.date())
    yesterday = weekday_name(local.date() - timedelta(days=1))
    for zone in zones:
        days = set(zone.get("days", WEEKDAYS))
        f_hh, f_mm = parse_hhmm(zone["from"])
        t_hh, t_mm = parse_hhmm(zone["to"])
        frm, to = time(f_hh, f_mm), time(t_hh, t_mm)
        if frm < to:
            if today in days and frm <= t < to:
                return True
        else:  # wraps midnight
            if (today in days and t >= frm) or (yesterday in days and t < to):
                return True
    return False


def utc_day_of(epoch: int) -> date:
    return datetime.fromtimestamp(epoch, UTC).date()


def utc_midnight(day: date) -> int:
    return int(datetime.combine(day, time(0, 0), tzinfo=UTC).timestamp())
